src/Utils/ProtoHelpers.php

```diff
@@ -105,6 +109,84 @@ public static function routingParameters(ProtoCatalog $catalog, ?DescriptorProto
+    {
+        if ($method->restRoutingHeaders && !$method->routingParameters) {
+            return static::implicitParamsDescriptor($method->restRoutingHeaders ?? Map::new());
+        }
+
+        return static::routingParamsDescriptor($method->routingParameters ?? Map::new());
+    }
+
+    /**
+     * Processes the google.api.http based implicit header request parameters of a method.
+     * 
+     * @param Map $implicitParams mapping of key -> Vector of strings, that being the field accessors.
+     * 
+     * @return array
+     */
+    public static function implicitParamsDescriptor(Map $implicitParams): array
+    {
+        return $implicitParams->keys()
+            ->map(fn($key) => [
+                'fieldAccessors' => $implicitParams->get($key, Vector::new())->toArray(),
+                'keyName' => $key
+            ])
+            ->toArray();
+    }
+
+    /**
+     * Processes the RoutingParameters (pre-processed by ProtoHelpers::routingParameters) into a descriptor
+     * config-friendly array for generation.
+     *
+     * @param Map $routingParameters output of ProtoHelpers::routingParameters.
+     *
+     * @return array all header param configurations for the method.
+     */
+    public static function routingParamsDescriptor(Map $routingParameters): array
+    {
+        $headerParams = [];
+
+        foreach($routingParameters->values() as $fieldRoutingParams) {
+            // Group by header key so that GAX the relevant matchers can be grouped in order.
+            $paramsByKey = $fieldRoutingParams->groupBy(fn($f) => $f['key']);
+            
+            // Group the matchers by header key in order.
+            $matchersByKey = $paramsByKey->mapValues(
+                fn($key, $fieldRoutingParams) =>
+                    $fieldRoutingParams
+                        ->filter(fn($f) => isset($f['regex']))
+                        ->map(fn($f) => $f['regex']))
+                ->filter(fn($key, $matchers) => $matchers)
+                ->mapValues(fn($key, $val) => $val->toArray());
+            
+            // Flatten the entire set of key-oriented configs, taking the first "getter" chain (should all be the same),
+            // and injecting the matchers for that header key if configured.
+            $flattened = $paramsByKey->mapValues(fn($key, $params) =>
+                // Only set the `matchers` if it is set and not empty.
+                isset($matchersByKey[$key]) && $matchersByKey[$key] ? [
+                    'fieldAccessors' => $params[0]['getter'],
+                    'keyName' => $key,
+                    'matchers' => $matchersByKey[$key]
+                ] : [
+                    'fieldAccessors' => $params[0]['getter'],
+                    'keyName' => $key
+                ]
+            )->values()->toArray();
+
+            $headerParams = array_merge($headerParams, $flattened);
+        }
+
+        return $headerParams;
+    }
+
```

Comment:
Most of this will be refactored when we rip out the header param injection code while moving to `startApiCall`. A deep review is probably not necessary.